src/2024h2/Contracts-and-invariants.md

```diff
@@ -0,0 +1,217 @@
+# Contracts and Invariants
+
+| Metadata | |
+| --- | --- |
+| Owner(s) | pnkfelix |
+| Teams | Lang, Libs, Compiler |
+| Status | WIP |
+
+## Motivation
+
+We wish to extend Rust in three ways:
+First, we want to extend the Rust language to enable Rust developers to write predicates, called "contracts", and attach them to specific points in their program. We intend for this feature to be available to all Rust code, including that of the standard library.
+Second, we want to extend the Rust crate format such that the contracts for a crate can be embedded and then later extracted by third-party tools.
+Third, we want to extend project-supported Rust compiler and interpreter tools, such as `rustc` and `miri`, to compile the code in a mode that dynamically checks the associated contracts (note that such dynamic checking might be forced to be incomplete for some contracts).
+Examples of contracts we envision include: pre- and post-conditions on Rust methods; representation invariants for abstract data types; and loop invariants on `for` and `while` loops. 
+
+Our motivation for this is that contracts are key for reasoning about software correctness. Formal verification tools such as [Creusot][], [Kani][], and [Verus][] have demonstrated that it is possible to write Rust code that is coupled with automated verification mechanisms. But furthermore, we assert that contracts can provide value even if we restrict our attention to dynamic checking: By having a dedicated construct for writing method specifications and invariants formally, we will give our tools new avenues for testing our programs in an automated fashion, similar to the benefits provdied by fuzzing.
+
+[Creusot]: https://github.com/creusot-rs/creusot?tab=readme-ov-file#about
+[Kani]: https://model-checking.github.io/kani/
+[Verus]: https://verus-lang.github.io/verus/guide/
+
+### The status quo
+
+Currently, if you want to specify the behavior of a Rust method and check that the specification is correct, you can either attempt construct a test suite that covers the entirety of your specification, or you can manually embed contract-like predicates into your code. Embedding contract-like predicates is typically done via variations of either 1. `assert!`, 2. `debug_assert!`, or 3. similar `cfg`-guarded code sequences that abort/panic when a predicate fails to hold.
```

Comment:
> you can either attempt construct a test suite that covers the entirety of your specification

Are you encompassing property-based testing here? It has interesting aspects that could be discussed for dynamic validation, e.g. not just what component broke the contract, but minimizing the steps needed to trigger the problem.

Reply:
Right: I view property-based testing as a way to describe many tests in a concise fashion. In some cases, one can indeed encode the entirety of a specification as a collection of property-based tests (just as, in some cases, one can encode all input/ouput pairs in an exhaustive test suite).

I think its actually likely for the contracts I envision to be used in a manner that is quite similar to property-based testing, though I am hoping to allow for a bit more expressiveness, in order to help out automated verification tools.

Does that help answer the question?

I might need more guidance about what you are referring to in terms of "minimizing steps needed to trigger the problem" (I can imagine various ways in which property-based testing achieves that, especially if we are talking about worlds where one adds new property-based tests during the development cycle, but I figure its better to just ask you for more info thatI can read.)

Reply:
Ah I meant "minimizing the steps needed to trigger the problem" literally: the test case reduction/shrinking procedure the PBT tool will apply to data that violates one of your properties. Like reducing the number of items in a vec, or some reified steps/actions/messages/whatever in model-based property tests.